pkg/helm/nelm/nelm.go

```diff
@@ -0,0 +1,254 @@
+		opts.KubeContext = *getter.Context
+	}
+
+	if opts.HelmDriver == "" {
+		opts.HelmDriver = os.Getenv("HELM_DRIVER")
+	}
+
+	// FIXME(ilya-lesikov): this function does A LOT. And with what addon-operator expects from us
+	// in regards to logging, all of this is kind of a mess now
+	opts.context = action.SetupLogging(context.TODO(), "info", "", "off")
+
+	commonOptions = opts
+
+	versionResult, err := action.Version(commonOptions.context, action.VersionOptions{
+		OutputNoPrint: true,
+	})
+	if err != nil {
+		return fmt.Errorf("get nelm version: %w", err)
+	}
+
+	logger.Info("Nelm version", slog.String("version", versionResult.FullVersion))
+
+	return nil
+}
+
+type CommonOptions struct {
+	Namespace  string
+	HistoryMax int32
+	// XXX: no global timeout in Nelm
+	Timeout     time.Duration
+	HelmDriver  string
+	KubeContext string
+	// XXX: add all other options from *genericclioptions.ConfigFlags
+
+	// FIXME(ilya-lesikov): we need a special context with logboek attached to it to be passed to
+	// Nelm actions to actually log anything
+	context context.Context
+}
+
+func NewNelmClient(logger *log.Logger, labels map[string]string) *NelmClient {
+	logEntry := logger.With("operator.component", "nelm")
+
+	return &NelmClient{
+		logger: logEntry,
+		labels: labels,
+	}
+}
+
+type NelmClient struct {
+	// FIXME(ilya-lesikov): not gonna be used in Nelm atm
+	logger *log.Logger
+	labels map[string]string
+}
+
+func (c *NelmClient) WithLogLabels(logLabels map[string]string) {
+	c.logger = utils.EnrichLoggerWithLabels(c.logger, logLabels)
+}
+
+func (c *NelmClient) WithExtraLabels(labels map[string]string) {
+	for k, v := range labels {
+		c.labels[k] = v
+	}
+}
+
+func (c *NelmClient) LastReleaseStatus(releaseName string) (string, string, error) {
+	releaseGetResult, err := action.ReleaseGet(commonOptions.context, releaseName, commonOptions.Namespace, action.ReleaseGetOptions{
+		KubeContext:          commonOptions.KubeContext,
+		OutputNoPrint:        true,
+		ReleaseStorageDriver: commonOptions.HelmDriver,
+	})
+	if err != nil {
+		// FIXME(ilya-lesikov): return a specific error from ReleaseGet if release not found
+		if regexp.MustCompile(`.+release ".+" (namespace ".+") not found`).Match([]byte(err.Error())) {
+			return "0", "", fmt.Errorf("get nelm release %q: %w", releaseName, err)
+		}
+
+		return "", "", fmt.Errorf("get nelm release %q: %w", releaseName, err)
+	}
+
+	return strconv.FormatInt(int64(releaseGetResult.Release.Revision), 10), releaseGetResult.Release.Status.String(), nil
+}
+
+func (c *NelmClient) UpgradeRelease(releaseName string, chartName string, valuesPaths []string, setValues []string, labels map[string]string, namespace string) error {
+	c.logger.Info("Running nelm install for release",
+		slog.String("release", releaseName),
+		slog.String("chart", chartName),
+		slog.String("namespace", namespace))
+
+	// XXX: for some reason upgrade was called twice previously?
+	// XXX: why we always deployed with upg.SkipCRDs = true?
```

Comment:
We have another flow. We discovery CRDs at the startup and [install](https://github.com/flant/addon-operator/blob/b7c97010128922e066fed35e2e53e91834a250de/pkg/addon-operator/operator.go#L1362) it for all enabled modules before module render. It allows to avoid races between modules which are based on some CRDs, like monitoring

Reply:
Will provide an option to not install CRDs in the Nelm action.